core/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java

```diff
@@ -0,0 +1,220 @@
+        }
+        this.input = input;
+    }
+
+    /** Returns a path representing the ES keystore in the given config dir. */
+    static Path keystorePath(Path configDir) {
+        return configDir.resolve("elasticsearch.keystore");
+    }
+
+    /** Constructs a new keystore with the given password. */
+    static KeyStoreWrapper create(char[] password) throws Exception {
+        KeyStoreWrapper wrapper = new KeyStoreWrapper(password.length != 0, NEW_KEYSTORE_TYPE, NEW_KEYSTORE_SECRET_KEY_ALGO, null);
+        KeyStore keyStore = KeyStore.getInstance(NEW_KEYSTORE_TYPE);
+        keyStore.load(null, null);
+        wrapper.keystore.set(keyStore);
+        wrapper.keystorePassword.set(new KeyStore.PasswordProtection(password));
+        return wrapper;
+    }
+
+    /**
+     * Loads information about the Elasticsearch keystore from the provided config directory.
+     *
+     * {@link #loadKeystore(char[])} must be called before reading or writing any entries.
+     * Returns {@code null} if no keystore exists.
+     */
+    public static KeyStoreWrapper loadMetadata(Path configDir) throws IOException {
+        Path keystoreFile = keystorePath(configDir);
+        if (Files.exists(keystoreFile) == false) {
+            return null;
+        }
+        DataInputStream inputStream = new DataInputStream(Files.newInputStream(keystoreFile));
+        int format = inputStream.readInt();
+        if (format != FORMAT_VERSION) {
+            throw new IllegalStateException("Unknown keystore metadata format [" + format + "]");
+        }
+        boolean hasPassword = inputStream.readBoolean();
+        String type = inputStream.readUTF();
+        String secretKeyAlgo = inputStream.readUTF();
+        return new KeyStoreWrapper(hasPassword, type, secretKeyAlgo, inputStream);
+    }
+
+    /** Returns true iff {@link #loadKeystore(char[])} has been called. */
+    public boolean isLoaded() {
+        return keystore.get() != null;
+    }
+
+    /** Return true iff calling {@link #loadKeystore(char[])} requires a non-empty password. */
+    public boolean hasPassword() {
+        return hasPassword;
+    }
+
+    /** Loads the keystore this metadata wraps. This may only be called once. */
+    public void loadKeystore(char[] password) throws GeneralSecurityException, IOException {
+        this.keystore.set(KeyStore.getInstance(type));
+        try (InputStream in = input) {
+            keystore.get().load(in, password);
+        }
+
+        this.keystorePassword.set(new KeyStore.PasswordProtection(password));
+        Arrays.fill(password, '\0');
+
+        // convert keystore aliases enum into a set for easy lookup
+        Enumeration<String> aliases = keystore.get().aliases();
+        while (aliases.hasMoreElements()) {
+            settingNames.add(aliases.nextElement());
+        }
+    }
+
+    /** Write the keystore to the given config directory. */
+    void save(Path configDir) throws Exception {
```

Comment:
We don't. My point was code outside of core should not be calling save (hence it is package private). But it is moot since we do not give permissions to write to the config dir, so only the keystore tool (which doesn't run under SM) can write it.